server/service/schedule/schedule.go

```diff
@@ -313,3 +334,49 @@ func (s *Schedule) acquireLock() bool {
 	level.Debug(s.logger).Log("msg", "not the lock leader, skipping")
 	return false
 }
+
+func (s *Schedule) releaseLock() {
+	err := s.locker.Unlock(s.ctx, s.getLockName(), s.instanceID)
+	if err != nil {
+		level.Error(s.logger).Log("msg", "unlock failed", "err", err)
+		sentry.CaptureException(err)
+	}
+}
+
+// holdLock attempts to acquire a schedule lock. If it successfully acquires the lock, it starts a
+// goroutine that periodically extends the lock, and it returns `true` along with a
+// context.CancelFunc that will end the goroutine and release the lock. If it is unable to initially
+// acquire a lock, it returns `false, nil`. The maximum duration of the hold is two hours.
+func (s *Schedule) holdLock() (bool, context.CancelFunc) {
+	if ok := s.acquireLock(); !ok {
+		return false, nil
+	}
+
+	const MAX_HOLD = 2 * time.Hour
+	ctx, cancelFn := context.WithDeadline(s.ctx, time.Now().Add(MAX_HOLD))
+
+	go func() {
+		ticker := time.NewTicker(s.getSchedInterval() / 10 * 8)
```

Comment:
What does the `/ 10 * 8` mean?

Reply:
Oh, this is 80% of the interval. I guess my mind was too set on `* 0.8` pattern.

Reply:
Our comments crossed lol

Working with `time.Duration` values didn't lend itself easily to floats so I took the fractional approach